sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -0,0 +1,245 @@
+# Minimum spanning tree single linkage implementation for hdbscan
+# Authors: Leland McInnes, Steve Astels
+# License: 3-clause BSD
+
+import numpy as np
+cimport numpy as cnp
+import cython
+
+from libc.float cimport DBL_MAX
+
+from sklearn.metrics._dist_metrics cimport DistanceMetric
+
+
+cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_linkage_core(
```

Comment:
Since there is no documentation, I had a bit of trouble figuring out what these things are doing.

Actually, I think that both `mst_linkage_core` and `mst_linkage_core_vector` are relying on the same algorithm.

The difference is that the former one relies on the fact that we computed all pairwise distances while the latter will compute the distance within the loop.

So several questions here:

- Is the algorithm used here the Prism's algorithm?
- From the paper, it seems that Boruvska is much more efficient but I think only with the KDTree and BallTree. So do we have to rely on Prism's algorithm for the brute force approach?
- Since `mst_linkage_core` and `mst_linkage_core_vector` are based on the same algorithm, I don't know how much flexibility we should provide to our user. From what I get from the paper, passing the distance matrix will provide a reduced computation complexity `O(n_samples ** 2)` but an increased memory of complexity of `O(n_samples ** 2)` since computing the distances on the go will have a memory complexity of `O(n_samples * n_features)` and computational complexity of `O(n_samples ** 2 * n_features)`. Since the memory-efficient version is used only with the KDTree and BallTree, I am wondering if we really need to implement it since in this case, we should rely on Boruvska's algorithm which is actually the method that scales (I mean the dual-tree traversal).

Reply:
1. Yes

2. Regarding Boruvka, originally it was implemented in Cython directly leveraging the `{KD, Ball}Tree` classes (hence not for `brute`) and added _a ton_ of code complexity to the project, so I've set that aside as possible functionality to reintroduce in a follow-up PR. We would also need to run some benchmarks to figure out when Boruvka vs Prims is preferable. We could reasonably implement a Boruvka brute-force approach, but that would obviously require benchmarking to compare as well. I think we should forego Boruvka for the initial release.

3. The original implementation had certain heuristics in place (namely number of features) to determine which algorithms (prims vs boruvka) to defer to. We've got several degrees of freedom here regarding implementation. Specifically we have:
- `distance_matrix` vs `data_matrix` computation (referring to `mst_linkage_core` and `mst_linkage_core_vector` vector resp.) in Prims.
- Prims vs Boruvka

I don't have any benchmarks currently, but I think we should revisit this after an initial implementation is merged into `main`. What do you think?

Reply:
The dual tree Boruvka implementation is decidedly complicated. Deferring that to another PR would be a sensible step to make reviewing tractable. On the other hand for any dimensionality where the ``{KD,Ball}Tree`` are efficient (less than 50 was a useful rule of thumb) compared to brute force is is *vastly* more efficient (approx ``O(N log N)`` instead of ``O(N^2)``) so you really want to have it eventually.

With regard to ``mst_linkage_core`` vs ``mst_linkage_core_vector`` I think some renaming may be useful, but both are valuable. You will need ``mst_linkage_core`` whenever you have ``metric="precomputed"`` which is a not uncommon use case, and it provides a good simple implementation for small data that you can test the other implementations against. On the other hand the memory complexity of this makes it intractable for large data, so ``mst_linkage_core_vector`` should definitely be the default choice if you have vector data (as opposed to just a distance matrix).